### PR TITLE
extra types

### PR DESCRIPTION
**Pull request by perceptron8**
This PR adds:
- `LocalDateRange`,
- `OffsetDate`,
- `Quarter`,
- `YearQuarter`,
- `YearWeek`.

It was necessary to modify `ChronoField.EPOCH_DAY.range()` (which was incorrect), otherwise `LocalDate.MIN.toEpochDay()` and `LocalDate.MAX.toEpochDay()` was outside of it.

It was also neccessary to mark `ChronoField.PROLEPTIC_MONTH` to be date-based as `OffsetDate` tests depend on that.

I wasn't sure how to include these 2 changes into compilation process with no ability to publish core, so locally I used `npm link ../core`. It's possible that CI build will fail. What is the preferred way to deal with such inter-module dependencies?

In 2 places I've used generator functions to imitate streams, so @babel/plugin-transform-regenerator is now required.

### CODE REVIEW

**Comment by perceptron8** [comment]
I'm stuck and need some help. Should @babel/plugin-transform-runtime be used instead of @babel/plugin-transform-regenerator to avoid `regeneratorRuntime` dependency?

**Comment by pithu** [comment]
> I wasn't sure how to include these 2 changes into compilation process with no ability to publish core, so locally I used npm link ../core. It's possible that CI build will fail. What is the preferred way to deal with such inter-module dependencies?

As an alternative to `npm link` you can use `lerna link` from root directory. It sym links all packages together

**Comment by pithu** [comment]
> In 2 places I've used generator functions to imitate streams, so @babel/plugin-transform-regenerator is now required.

Should we really implement methods returning the `Stream` interface ? Since we do not have a replacement for the `Stream` interface in javascript, the API would be different. Or is there an alternative for returning a `generator` ? What about keep it simple and return an array ?

**Comment by perceptron8** [comment]
I wouldn't consider an array to be good replacement of the `Stream` as it's not lazy. Iterating over `LocalDateRange.ALL.stream()` would be unpractical or even impossible because of memory requirements. On the other hand, generators implement `Iterable` protocol, so one can simply use `for..of`. It's also very easy to get an array from generator / iterable by using `Array.from()` (or even spread operator in case of smallest ranges).

In my opinion, we do not need exact stream substitute. There's no parallelism, so we can forget about `spliterator()`, `parallel()`, `close()` and others. We don't have `Collectors` too, but if there is some great "collecting" library out there, it's probably using generators / iterables as it's main, or at least supported abstraction. For instance, please note that `Iterable` is listed in rxjs [from](https://rxjs.dev/api/index/function/from)#([ObservableInput](https://rxjs.dev/api/index/type-alias/ObservableInput)).

We can also remove methods returning the `Stream` for now, as you suggested, and wait if someone asks for them.

**Comment by pithu** [comment]
I agree that arrays are not a good replacement for generators / iterables for larger arrays. On the the other hand i am not sure what are the real world use cases for LocalDateRange and YearQuarter. And lot of questions arise in this context: eg Are there more cases where we would need streams. are generator the best replacement, what are the typical use cases, what libraries are out there in javascript that consume streams, should we even transform generators or just leave the native implementation (most browser except ie11 should be compatible), etc ?

So before implementing "java stream methods" we should answer all this questions and have a solid and fix implementation for that cases, that we will not have to change in future.

So for now i would suggest to split out the two "stream methods" and save for later. WDYT ?

**Comment by perceptron8** [comment]
> So for now i would suggest to split out the two "stream methods" and save for later. WDYT ?

Done!

**Comment by perceptron8** [comment]
Sorry, it's still WIP.

**Comment by perceptron8** [comment]
Now it should be ready.

File: packages/extra/src/OffsetDate.js [hunk 1]
```diff
@@ -0,0 +1,1274 @@
...
+     * This will query the {@link Clock.systemDefaultZone} system clock in the default
+     * time-zone to obtain the current date.
+     * The offset will be calculated from the time-zone in the clock.
+     * 
+     * Using this method will prevent the ability to use an alternate clock for testing
+     * because the clock is hard-coded.
+     *
+     * @return {OffsetDate} the current date using the system clock, not null
+     */
+    static now0() {
```
Comment (pithu) [inline packages/extra/src/OffsetDate.js:70]: why we don't have a leading underscore here ? do we have to add @private annotation, same for the other private methods

File: packages/extra/src/OffsetDate.js [hunk 2]
```diff
@@ -0,0 +1,1274 @@
...
+        this._offset = requireNonNull(offset, 'offset');
+    }
+
+    /**
+     * Returns a new date based on this one, returning `this` where possible.
+     *
+     * @param {LocalDate} date  the date to create with, not null
+     * @param {ZoneOffset} offset  the zone offset to create with, not null
+     */
+    _with(date, offset) {
```
Comment (pithu) [inline packages/extra/src/OffsetDate.js:248]: Why to we have a `_with` but no implementation for the overide of `with(...)` ? Do i oversee something ?

**Review by pithu — COMMENTED** [review]
Amazing work! Thanks a lot. My main comment here is, if we could unify the naming / declaration of private methods

File: packages/extra/src/LocalDateRange.js [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+     * @returns {LocalDateRange}
+     */
+    static of(startInclusive, endExclusiveOrPeriod) {
+        if (startInclusive instanceof LocalDate && endExclusiveOrPeriod instanceof LocalDate) {
+            return LocalDateRange._ofLocalDateLocalDate(startInclusive, endExclusiveOrPeriod);
+        }
+        if (startInclusive instanceof LocalDate && endExclusiveOrPeriod instanceof Period) {
+            return LocalDateRange._ofLocalDatePeriod(startInclusive, endExclusiveOrPeriod);
+        }
+        // FIXME
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:61]: Oh, that's indeed something I forgot to fix. If you invoke `of()` passing (`LocalDate`, `Foo`), you will get _"startInclusive must be an instance of LocalDate but is LocalDate ..."_. Similarly, if you pass (`Foo`, `LocalDate`), you will get _"... endExclusiveOrPeriod must be an instance of LocalDate or Period but is LocalDate"_. Finally, for (`Foo`, `Period`), you will get _"... endExclusiveOrPeriod must be an instance of LocalDate or Period but is Period"_. Give me some time, please.

File: packages/extra/src/OffsetDate.js (hunk 1 above)
Comment (perceptron8) [reply packages/extra/src/OffsetDate.js:70]: I used `YearMonth` as reference, however I agree that making this method private would be nice.

File: packages/extra/src/OffsetDate.js (hunk 2 above)
Comment (perceptron8) [reply packages/extra/src/OffsetDate.js:248]: `OffsetDate._with` is not related to `Temporal.with`. See https://github.com/ThreeTen/threeten-extra/blob/v1.7.1/src/main/java/org/threeten/extra/OffsetDate.java#L327.

Underscore needs to be here because 1. this method is not public, 2. even if not, it cant be named `with` because it would overload `Temporal.with`.

Same goes for `YearQuarter._with`.

File: packages/extra/src/LocalDateRange.js [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+     * No empty range can exist at `LocalDate.MIN` or `LocalDate.MAX`.
+     *
+     * @param {LocalDate} startInclusive  the inclusive start date, not null
+     * @param {LocalDate} endExclusive  the exclusive end date, not null
+     * @return {LocalDateRange} the half-open range, not null
+     * @throws DateTimeException if the end is before the start,
+     *   or the start date is `LocalDate.MAX` or `LocalDate.MAX.minusDays(1)`,
+     *   or the end date is `LocalDate.MIN` or `LocalDate.MIN.plusDays(1)`
+     */
+    static _ofLocalDateLocalDate(startInclusive, endExclusive) {
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:89]: What do you mean by "excluded from the docs"?

As for `@Private`, I would be more than happy to change almost all `_` to [`#`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) - probably with the exception of protected-like methods needed by `Temporal`.

In the meanwhile, I'll try to find if it's possible in JS to use template method patterns and support encapsulation at the same time.

File: packages/extra/src/OffsetDate.js (hunk 1 above)
Comment (perceptron8) [reply packages/extra/src/OffsetDate.js:70]: `Year` and `MonthDay` also have `now0()` with no underscore. Should I fix them too?

**Comment by InExtremaRes** [comment]
Hi again. I've been super-very short of time lately, that's why I'm been very absent, but I will try to take a look at least to the TypeScript part of this PR between today and tomorrow.

File: packages/extra/src/LocalDateRange.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js]: Fixed in 6dc4426.

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/src/LocalDateRange.js:93]: Its more like a convention. all internal methods, that should not be used from "outside" and are not part of the API, should start with an `_` and have the `@private` annotation. The `@private` annotation excludes it from the generated esdocs. The `_` makes it obvious for developers, that are inspecting the code.
I am not sure if we fulfill this convention everywhere in our code base, but the most places in our `core` should.

File: packages/extra/src/OffsetDate.js (hunk 1 above)
Comment (pithu) [reply packages/extra/src/OffsetDate.js:70]: Yes please as mentioned above, all internal methods should start with an `_` and have the `@private` annotation

Comment (perceptron8) [reply packages/extra/src/OffsetDate.js:70]: My impression is that if `now0()` and others becomes private, jsdoc of `now()` is going to be useless. See:
- https://js-joda.github.io/js-joda/class/packages/core/src/Year.js~Year.html#static-method-now
- https://js-joda.github.io/js-joda/class/packages/core/src/YearMonth.js~YearMonth.html#static-method-now
- https://js-joda.github.io/js-joda/class/packages/core/src/MonthDay.js~MonthDay.html#static-method-now

`now()` jsdoc points to `now0()`, `nowClock(clock: Clock)` and `nowZoneId(zone: ZoneId)`. There is similar pattern in case of several `of()` methods. And that's not even the main problem! **These methods are public for at least 6 years**. I'm not sure if making them private now is going to make users happy.

**Review by InExtremaRes — CHANGES_REQUESTED** [review]
Thank you very much for your work! Most of the TypeScript definitions seem fine to me. I left notes to some small issues, like classes lacking a private constructor.

Hope your are doing well.

File: packages/extra/src/OffsetDate.js (hunk 1 above)
Comment (InExtremaRes) [reply packages/extra/src/OffsetDate.js:70]: I agree with @pithu here. Those methods should be private, and should have been since the beginning. What you are saying about the documentation is true though, but I think that's something that should be fixed at a larger scale. A lot of the documentation needs improvements and the issue you're pointing out is an issue everywhere in this project. Also renaming those methods to start with `_` in other classes (in `core` for instance) is of course a breaking change, but these classes are new; I think is better to make things right from the start here.

However, that being said, I'm an advocate to clean the interfaces in other packages and classes as well, renaming all private methods even if it's a breaking change. Those methods are implementation details and we should be able to change them or delete them at any time.

In this case I would prefer that every method that is not part of the public API is named with an `_` and marked as `@Private`.

Comment (perceptron8) [reply packages/extra/src/OffsetDate.js:70]: > Those methods should be private, and should have been since the beginning.

Sure! I do agree with both of you and was even willing (and still am) to go further and use true [private class methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#private_methods) instead of convention in whole extra package.

> I'm an advocate to clean the interfaces in other packages and classes as well, renaming all private methods even if it's a breaking change.

+1 for cleaning core interfaces sometime later. For now, `@deprecated` could be helpful.

> In this case I would prefer that every method that is not part of the public API is named with an _ and marked as `@Private`.

There are methods (e.g. Temporal's _withField, _withAdjuster, _plusUnit, _plusAmount) which seems to be more `@protected`-like. Still non-public, but with clearly marked purpose.

Comment (InExtremaRes) [reply packages/extra/src/OffsetDate.js:70]: @perceptron8

Using `@deprecated` is a very good suggestion.

> There are methods (e.g. Temporal's _withField, _withAdjuster, _plusUnit, _plusAmount) which seems to be more `@protected`-like. Still non-public, but with clearly marked purpose.

Yes, that's exactly right. We changed (or added) those methods to unify some bits of implementation details and to clean the inheritance chain not so long ago. They are in fact protected in spirit (and in the types definitions), meant to be used by anyone implementing a `Temporal` type. We used `_` as well since they shouldn't be used by anyone outside the inheritance chain. We don't have any established convention right now to differentiate this protected-like methods from private ones, so we used `_` for anything non-public.

About using private fields I personally wouldn't mind. Native support is increasing in every important platform. My only concern is that js-joda still targets ES5, which doesn't include `WeakMap`. How does Babel transpile them to ES5? If it does so by prepending the name with and underscore or something I'd prefer just going as we are for now. Besides, I'm not sure if the performance could degrade if it transpiles to a `WeakMap` in ES6+. 
I think it is worth to consider and discuss the use of private fields/methods, though.

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:93]: As for docs, I'm unable to force JSDoc to document more than one overload.

```javascript
/**
 * (...)
 * @param {LocalDate} date (...)
 * @param {ZoneOffset} offset (...)
 * @return {OffsetDate} (...)
 *//**
 * (...)
 * @param {number} year (...)
 * @param {number} month (...)
 * @param {number} dayOfMonth (...)
 * @param {ZoneOffset} offset (...)
 * @return {OffsetDate} (...)
 * @throws DateTimeException (...)
 */
static of() {
    switch (arguments.length) {
        case 2: return OffsetDate._ofLocalDateZoneOffset(...arguments);
        case 4: return OffsetDate._ofIntIntIntZoneOffset(...arguments);
        default: throw new IllegalArgumentException('Illegal number of arguments');
    }
}
```

Building esdoc ends up with:
```console
warning: signature mismatch: OffsetDate.of
```

It's also worth nothing, that with this ugly pattern there is no way to document `IllegalArgumentException` thrown from dispatching method itself.
On the other hand, making `OffsetDate._of*` public just for the sake of documentation doesn't feel right.

@pithu, @InExtremaRes Any ideas?

File: packages/extra/typings/js-joda-extra.d.ts [hunk 5]
```diff
@@ -79,5 +79,240 @@ export class Interval {
...
+    equals(obj: any): boolean;
+    end(): LocalDate;
+    endInclusive(): LocalDate;
+    start(): LocalDate;
+    hashCode(): number;
+    intersection(other: LocalDateRange): LocalDateRange;
+    isAfter(date: LocalDate): boolean;
+    isAfter(other: LocalDateRange): boolean;
+    isBefore(date: LocalDate): boolean;
+    isBefore(other: LocalDateRange): boolean;
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts]: Done in fb39ecc.

File: packages/extra/typings/js-joda-extra.d.ts [hunk 6]
```diff
@@ -79,5 +79,240 @@ export class Interval {
...
+    isConnected(other: LocalDateRange): boolean;
+    isEmpty(): boolean;
+    isUnboundedEnd(): boolean;
+    isUnboundedStart(): boolean;
+    lengthInDays(): number;
+    overlaps(other: LocalDateRange): boolean;
+    span(other: LocalDateRange): LocalDateRange;
+    stream(): Generator<LocalDate>;
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts]: Done in 521c32e.

File: packages/extra/typings/js-joda-extra.d.ts [hunk 7]
```diff
@@ -79,5 +79,240 @@ export class Interval {
     toString(): string;
 }
 
+export class LocalDateRange {
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts:82]: Done in 694b438.

File: packages/extra/typings/js-joda-extra.d.ts [hunk 8]
```diff
@@ -79,5 +79,240 @@ export class Interval {
...
+    span(other: LocalDateRange): LocalDateRange;
+    stream(): Generator<LocalDate>;
+    toPeriod(): Period;
+    toString(): string;
+    union(other: LocalDateRange): LocalDateRange;
+    withEnd(adjuster: TemporalAdjuster): LocalDateRange;
+    withStart(adjuster: TemporalAdjuster): LocalDateRange;
+}
+
+export class OffsetDate extends Temporal implements TemporalAdjuster {
```
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts:119]: Also done in 694b438.

File: packages/extra/typings/js-joda-extra.d.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts]: Done in fb39ecc.

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:93]: I just realized JSDoc isn't working properly because... it's not used at all.
Are there any plans for migrating to it from already dead ESDoc?

File: packages/extra/typings/js-joda-extra.d.ts [hunk 9]
```diff
@@ -79,5 +79,240 @@ export class Interval {
...
+
+    protected _minusUnit(amountToSubtract: number, unit: TemporalUnit): YearWeek;
+    protected _minusAmount(amount: TemporalAmount): YearWeek;
+    protected _plusUnit(amountToAdd: number, unit: TemporalUnit): YearWeek;
+    protected _plusAmount(amount: TemporalAmount): YearWeek;
+    protected _withAdjuster(adjuster: TemporalAdjuster): YearWeek;
+    protected _withField(field: TemporalField, newValue: number): YearWeek;
+}
+
+declare class Locale {
```
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/typings/js-joda-extra.d.ts:315]: agree

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/src/LocalDateRange.js:93]: No, we don't have any plans to migrate away from esdoc.

File: packages/extra/typings/js-joda-extra.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts:315]: ... or should I remove the whole `Quarter#displayName()` declaration from typings?

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/src/LocalDateRange.js:93]: Running `npm run build-esdoc` works just fine for me, so i might not understand your issue.

I see that there are still some `now0()` methods not declared private. What about change the name to `_now0()` ?

Beside the issue with Locale mentioned by @InExtremaRes, all the rest looks good to me now.

[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:93]: This makes documenting of`OffsetDate#of()` virtually impossible. Either `OffsetDate._of*`have to be public (which is ugly) or ESDoc supports overloads (which won't happen).

File: packages/extra/typings/js-joda-extra.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/typings/js-joda-extra.d.ts]: For now it's removed entirely in 30546024213fffcd537c7d37c8aafc6502d8bad2.

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:93]: > Running `npm run build-esdoc` works just fine for me, so i might not understand your issue.

Either
Change `OffsetDate.of()` docs as in https://github.com/js-joda/js-joda/pull/622#discussion_r960829833. ESDoc doesn't support multiple `/** ... */`, so it will document just one of overload.

Or
Leave `OffsetDate.of()` as is, but make `ofLocalDateZoneOffset` and `ofIntIntIntZoneOffset` `@private`. This way documentation becomes useless, as there won't be anything to link to.

```
     * if called with 2 arguments {@link OffsetDate.ofLocalDateZoneOffset} is executed,
     * if called with 4 agruments {@link YearQuarter.ofIntIntIntZoneOffset} is executed,
```

BTW YearQuarter.ofIntIntIntZoneOffset should be changed to OffsetDate.ofIntIntIntZoneOffset.

File: packages/extra/typings/js-joda-extra.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/typings/js-joda-extra.d.ts]: :+1:

File: packages/extra/src/LocalDateRange.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/extra/src/LocalDateRange.js:93]: For me its fine like it is with `OffsetDate.of()`, the typescript typing showing the correct API and to have this "ugly" public methods like `ofIntIntIntZoneOffset` in the js API is not nice, but the limitation we have when migration from a language with function overloading

But why not fixing the `now0` method ? and some private constructors are still in the esdoc, these ones could be removed from esdoc with the `@private` annotation.

In summary, where possible we should have the docs aligned with the typescript definitions and where not possible we have to live with some work arounds.

[the remark this answers is not in the document]
Comment (perceptron8) [reply packages/extra/src/LocalDateRange.js:93]: > For me its fine like it is with `OffsetDate.of()`, the typescript typing showing the correct API and to have this "ugly" public methods like `ofIntIntIntZoneOffset` in the js API is not nice, but the limitation we have when migration from a language with function overloading.

It's just because of ESDoc. We could have no "ugly" public methods at all. Maybe migrating away from ESDoc is worth reconsidering? As a last resort I'll rename `ofIntIntIntZoneOffset` to `_ofIntIntIntZoneOffset`, but it still will have to public, probably forever.

> But i stumbled over some private constructors that are still in the esdoc, these ones could be removed from esdoc with the @private annotation. eg the OffsetDate constructor

Sure, fixed this already but didn't push because I was waiting for `of*()`-related discussion to be resolved.

**Comment by pithu** [comment]
@InExtremaRes do you have any further comments on this PR ?

**Comment by InExtremaRes** [comment]
TypeScript definitions look fine to me now. The only comment I have is about those methods like `now0` or `nowClock`. Was it the decision to name them with a leading `_`?

**Comment by perceptron8** [comment]
Doesn't have a chance to work because of https://github.com/js-joda/js-joda/blob/v2.2.1/esdoc.json#L12.

**Comment by perceptron8** [comment]
My bad. It _is_ possible for `_`* methods to be public if they are explicitly marked with `@public`. They are however displayed before all other methods because ESDoc appears to use alphabetical sorting.

**Comment by pithu** [comment]
> My bad. It _is_ possible for `_`* methods to be public if they are explicitly marked with `@public`. They are however displayed before all other methods because ESDoc appears to use alphabetical sorting.

Best option IMHO would be to have these private functions prefixed with `_`, marked as esdoc `@private`, but as well listed somehow in the esdoc generated docs. But i am not sure if this is possible.

**Comment by perceptron8** [comment]
> Best option IMHO would be to have these private methods prefixed with `_`, marked as esdoc `@private`, but as well listed somehow in the esdoc generated docs. But i am not sure if this is possible.

They still would be listed in the worst possible place: before any other methods. Best option? To hide `now0()` and other private methods and simply list all `now()` overloads one by one.

**Comment by pithu** [comment]
I just fiddled around a bit with esdoc. So including all private methods is a bad idea, because the docs get spammed with all this `_init` methods and other internal stuff.

What about marking the `private` methods that we want to include in the esdoc as `@protected` and prefix this methods with `_` ? And add `protected` in the esdoc config accessor list. still a bit weird because protected is uncommon in javascript. but wouldn't that be  the closest way to express what we want to say.

**Comment by pithu** [comment]
> They still would be listed in the worst possible place: before any other methods. Best option? To hide now0() and other private methods and simply list all now() overloads one by one.

With the `protected trick these methods are listed below the public methods in an own protected methods section

**Comment by perceptron8** [comment]
> With the `protected trick these methods are listed below the public methods in an own protected methods section

I've tried to use `@protected` trick before but got nothing as I knew nothing about esdoc config accessor list at the time. Thanks veru much for the hint! This seems to work really nice and is indeed closest to what we want to say (even if `static protected` is uncommon thing even outside js world ;)).

**Comment by perceptron8** [comment]
What's still missing is several `@param` and `@return` types. Found this just now.  I'll include them later today.

File: packages/extra/src/OffsetDate.js (hunk 1 above)
Comment (perceptron8) [reply packages/extra/src/OffsetDate.js]: I'm afraid that babel transpiles `#` to `WeakMap`, so for now let's stick to `_` convention.

As for ESDoc + overloading related problems, they were solved in likely best possible way thanks to @pithu's https://github.com/js-joda/js-joda/pull/622#issuecomment-1242922282.

**Comment by perceptron8** [comment]
Overloads are now fully documented, some `_`* methods became visible but not public. Also types are now present in every `@param` and `@return`. This PR is hopefully ready to be reviewed again.

**Comment by pithu** [comment]
Perfect! Great work, LGTM, i will do some final tests the next days and then merge it.

**Comment by perceptron8** [comment]
Sure! Take your time. And thanks for help (both of you!). If something comes up, let me know.

**Review by pithu — APPROVED** [review]
LGTM, thanks a lot
